python_modules/dagster/dagster/_core/definitions/decorators/assets_definition_factory.py

```diff
@@ -202,59 +205,115 @@ def make_keys_by_output_name(
+            "Cannot pass resource keys and decorate the function with resources",
+        )
+        return set(resource_defs_bound_to_asset.keys()) - arg_resource_keys
+    else:
+        return explicitly_passed_required_resource_keys | set(resource_defs_bound_to_asset.keys())
+
+
+class AssetsDefinitionBuilderArgs(NamedTuple):
+    name: Optional[str]
+    description: Optional[str]
+    specs: Sequence[AssetSpec]
+    check_specs: Sequence[AssetCheckSpec]
+    asset_out_map: Mapping[str, AssetOut]
+    upstream_asset_deps: Optional[Iterable[AssetDep]]
+    asset_deps: Mapping[str, Set[AssetKey]]
+    asset_in_map: Mapping[str, AssetIn]
+    can_subset: bool
+    group_name: Optional[str]
+    partitions_def: Optional[PartitionsDefinition]
+    retry_policy: Optional[RetryPolicy]
+    code_version: Optional[str]
+    op_tags: Optional[Mapping[str, Any]]
+    config_schema: Optional[UserConfigSchema]
+    retry_policy: Optional[RetryPolicy]
+    compute_kind: Optional[str]
+    required_resource_keys: Set[str]
+    resource_defs: Mapping[str, ResourceDefinition]
+    backfill_policy: Optional[BackfillPolicy]
+
+
+class AssetsDefinitionBuilder:
     def __init__(
         self,
         *,
-        directly_passed_asset_ins: Mapping[str, AssetIn],
         input_tuples_by_asset_key: Mapping[AssetKey, Tuple[str, In]],
         output_tables_by_asset_key: Mapping[AssetKey, Tuple[str, Out]],
-        check_specs: Sequence[AssetCheckSpec],
         internal_deps: Mapping[AssetKey, Set[AssetKey]],
-        can_subset: bool,
-        deps_directly_passed_to_multi_asset: Optional[Iterable[AssetDep]],
-        spec_resolver: Callable[["InOutMapper"], Sequence[AssetSpec]],
         op_name: str,
-        group_name: Optional[str] = None,
+        multi_asset_args: AssetsDefinitionBuilderArgs,
+        fn: Callable[..., Any],
     ) -> None:
-        self.directly_passed_asset_ins = directly_passed_asset_ins
         self._passed_input_tuples_by_asset_key = input_tuples_by_asset_key
         self.output_tuples_by_asset_key = output_tables_by_asset_key
-        self.check_specs = check_specs
         self.internal_deps = internal_deps
-        self.can_subset = can_subset
-        self.deps_directly_passed_to_multi_asset = deps_directly_passed_to_multi_asset
-        self.spec_resolver = spec_resolver
         self.op_name = op_name
-        self.group_name = group_name
+        self.args = multi_asset_args
+        self.fn = fn
+
+    @staticmethod
+    def from_args(*, fn: Callable[..., Any], args: AssetsDefinitionBuilderArgs):
+        op_name = args.name or fn.__name__
+
+        if args.asset_out_map and args.specs:
+            raise DagsterInvalidDefinitionError("Must specify only outs or specs but not both.")
+
+        if args.specs:
+            if args.upstream_asset_deps:
+                raise DagsterInvalidDefinitionError(
+                    "Can not pass deps and specs to @multi_asset, specify deps on the AssetSpecs"
```

Comment:
Similar comment here to above about this assuming it's called from the `@multi_asset` decorator.

Reply:
Pretty sure this is only hit in the multi_asset case. Will add invariant to confirm.

Reply:
https://app.graphite.dev/github/pr/dagster-io/dagster/22232/Have-asset-use-AssetsDefinitionBuilder#comment-PRRC_kwDOB9hbPs5hKTNA